### DISCUSSION TITLE
Testing Infrastructure that is easy to use in IDEs other than IntelliJ

### DISCUSSION THREAD

**Discussion by tkvangorder** [openrewrite/rewrite#304] [Ideas] [answered]
The testing infrastructure in rewrite-test is very Kotlin-centric and works well when using IntelliJ, however, this can make it more difficult for those attempting to evaluate Rewrite's capabilities when they are not using IntelliJ. 

I think with minimal effort we could provide a seamless testing experience for those wanting to use a pure-Java testing approach. 

- It should be easy to extend rewrite-test with java-based tools that do not rely on Kotlin.
- To work-around the lack of multi-line strings prior to Java 15, the testing infrastructure could provide utilities to load snippets from text files located on the classpath.

**Comment by sambsnyd** [comment]
My interpretation of this is that authoring tests in Java should be possible. Which means a plain old Java base for tests. I expect we can do this in a way that, probably via kotlin extension methods, none of our test classes would have to change.

**Comment by jkschneider** [comment]
`RewriteTest` needs to be in Kotlin for now, because only Kotlin methods can be called with named parameters, and our unit tests rely heavily on this:

```kotlin
assertChanged(
  before = """
    ...
  """,
  after = """
    ...
  """
)
```

**Comment by timtebeek** [comment, marked as answer]
There's now a [RewriteTest.java](https://github.com/openrewrite/rewrite/blob/main/rewrite-test/src/main/java/org/openrewrite/test/RewriteTest.java), which does not require Kotlin to write tests; We're transitioning to Java 17 for tests, to take advantage of text blocks, and slowly converting project tests from Kotlin to Java.

That ought to help developers setup IDEs other than IntelliJ, although [the recommended IDE remains IntelliJ](https://docs.openrewrite.org/authoring-recipes/recipe-development-environment), as [also indicated in CONTRIBUTORS.md](https://github.com/openrewrite/rewrite/blob/main/CONTRIBUTORS.md).
